### PR TITLE
enhancement(api): Support metric events in vector tap

### LINKED ISSUE #6892: Return metric events from tap
Currently, the tap API introduced in #6610 allows for returning log events, but ignores metrics.

We already have a `metric_to_log` transform that could be reused for tap. Just needs a bit of generalising to decouple from the transform API.

### PR DESCRIPTION
**Pull request by 001wwang**
Closes #6892

This PR adds basic support for tapping metric events via `vector tap`, introducing a new `Metric` object to the GraphQL API. There is likely more follow up that can be done to better model the `MetricValue` and extend functionality but I thought this made for a nice, small first pass.

TODO
- [x] Add tests

### Example
For the following config:
```toml
[api]
enabled = true

[sources.in]
type = "demo_logs"
format = "shuffle"
lines = [
  "test1",
  "test2",
]

[sources.internal_metrics]
type = "internal_metrics"

[sinks.out]
type = "blackhole"
inputs = ["in*"]
```

Running `vector tap "internal_metrics"`, you should see metrics in your console.
Screen Shot 2022-02-04 at 6 15 09 PM

You can also test out these changes in the GraphQL API playground with the following subscription:
```graphql
subscription {
  outputEventsByComponentIdPatterns(patterns: ["internal_metrics"]) {
    ... on Metric {
      componentId
      name
      namespace
      tags
      valueType
      value
      kind
      string(encoding: JSON)
    }
  }
}
```

which should display results such as 
```json
{
  "__typename": "Metric",
  "componentId": "internal_metrics",
  "name": "processed_events_total",
  "namespace": "vector",
  "tags": {
    "component_id": "to_metric",
    "component_kind": "transform",
    "component_name": "to_metric",
    "component_type": "log_to_metric",
    "output": "_default"
  },
  "valueType": "counter",
  "value": "224",
  "kind": "absolute",
  "string": "{\"name\":\"processed_events_total\",\"namespace\":\"vector\",\"tags\":{\"component_id\":\"to_metric\",\"component_kind\":\"transform\",\"component_name\":\"to_metric\",\"component_type\":\"log_to_metric\",\"output\":\"_default\"},\"timestamp\":\"2022-02-04T23:02:15.599170381Z\",\"kind\":\"absolute\",\"counter\":{\"value\":224.0}}"
}
```

### CODE REVIEW

File: lib/vector-api-client/src/gql/tap.rs [hunk 1]
```diff
@@ -51,11 +51,12 @@ impl From<TapEncodingFormat>
 }
 
 impl output_events_by_component_id_patterns_subscription::OutputEventsByComponentIdPatternsSubscriptionOutputEventsByComponentIdPatterns {
-    pub fn as_log(
+    pub fn as_string(
         &self,
-    ) -> Option<&output_events_by_component_id_patterns_subscription::OutputEventsByComponentIdPatternsSubscriptionOutputEventsByComponentIdPatternsOnLog>{
```
Comment (spencergilbert) [inline lib/vector-api-client/src/gql/tap.rs:56]: The name of this type pains me 😆

**Review by spencergilbert — APPROVED** [review]
Later will we be adding options to `tap` only logs or metrics?

**Comment by 001wwang** [comment]
> Later will we be adding options to `tap` only logs or metrics?

I don't think that's currently covered in #8292, but I can definitely see us adding an option to filter by event type.

**Comment by spencergilbert** [comment]
> > Later will we be adding options to `tap` only logs or metrics?
> 
> I don't think that's currently covered in #8292, but I can definitely see us adding an option to filter by event type.

Cool that seemed a natural extention at some point - granted you can _somewhat_ do that with the component selection

File: src/api/schema/events/metric.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        self.event.data().timestamp().as_ref()
+    }
+
+    /// Metric name
+    async fn name(&self) -> &str {
+        self.event.name()
+    }
+
+    /// Metric namespace
+    async fn namespace(&self) -> Option<&str> {
```
Comment (jszwedko) [inline src/api/schema/events/metric.rs:42]: I would have expected these fields (`namespace`, `kind`, etc.) to show up in the generated schema above, but they don't seem to be? Am I missing something? I do see the fields are actually returned by queries.

File: src/api/schema/events/metric.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        self.event.namespace()
+    }
+
+    /// Metric kind
+    async fn kind(&self) -> &str {
+        match self.event.kind() {
+            event::MetricKind::Incremental => "incremental",
+            event::MetricKind::Absolute => "absolute",
+        }
+    }
```
Comment (jszwedko) [inline src/api/schema/events/metric.rs:52]: We may want to make this a GraphQL enum type rather than string.

**Review by jszwedko — COMMENTED** [review]
Awesome! Just a couple of questions, but otherwise looks great!

**Comment by jszwedko** [comment]
Thanks also for providing the test configuration, commands, and expected output. That makes reviewing much easier.

File: src/api/schema/events/metric.rs (hunk 2 above)
Comment (jszwedko) [reply src/api/schema/events/metric.rs:42]: I actually do see them in your follow-up PR, which is odd: https://github.com/vectordotdev/vector/pull/11234/files#diff-65a33a582e1355cb4649213aceb696fe92a2339b16399c84137b511f696f90f8

File: src/api/schema/events/metric.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    /// Metric value in human readable form
+    async fn value(&self) -> String {
+        self.event.value().to_string()
+    }
+
+    /// Metric tags
+    async fn tags(&self) -> Option<&MetricTags> {
+        self.event.tags()
+    }
```
Comment (jszwedko) [inline src/api/schema/events/metric.rs:67]: Actually, I see now tags are rendered as an arbitrary set of key/value pairs in GraphQL. I think the more GraphQL-y approach would be to make `tags` a `[MetricTag!]` where `MetricTag` is

```graphql
type MetricTag {
  key: String!
  value: String!
}
```

File: src/api/schema/events/metric.rs (hunk 2 above)
Comment (001wwang) [reply src/api/schema/events/metric.rs:42]: Ah good catch, I think I may have missed re-running `make build-graphql-schema` at some point on this branch. Will update this.

File: src/api/schema/events/metric.rs (hunk 3 above)
Comment (001wwang) [reply src/api/schema/events/metric.rs]: Changed in [d4eed34](https://github.com/vectordotdev/vector/pull/11201/commits/d4eed349868db5cba309a8dc49c7e874d95a121d)

File: src/api/schema/events/metric.rs (hunk 4 above)
Comment (001wwang) [reply src/api/schema/events/metric.rs]: Updated in [ea4c581](https://github.com/vectordotdev/vector/pull/11201/commits/ea4c5818e1c8ff47780c109a30d3280a8af6da6a)

**Review by jszwedko — APPROVED** [review]
Looks like there are some unresolved imports, otherwise looks good to me! Thanks for the updates!

**Pull request merged**